### ISSUE TITLE
CUDA error: device-side assert triggered

### ISSUE THREAD

**Issue by gergo7633** [MIC-DKFZ/nnUNet#1419]
Dear Fabian,

I have an annoying stochastic error related to device-side assert.

The labels are good, I've checked them all by hand and also verified by nnUNET's --verify_dataset_integrity.

Usually 2 out of the 5 folds gets terminated by this error at various epochs (25, 73, 162, 700, etc).

I can continue them with the --c from the last checkpoint and they have finished without any further error.

I tried to set the CUDA_LAUNCH_BLOCKING=1 env variable to get a better trace, but obviously nnUNET never crashed those times. I'll set this to permanent for future errors.

I think it is not related to the labels, neither the number of classes (I have only background (0) and metastasis (1)).

Can this be a result of the loss function having a value outside the valid range somehow? 

[side_assert error.txt](https://github.com/MIC-DKFZ/nnUNet/files/11350888/side_assert.error.txt)
[side_assert error2.txt](https://github.com/MIC-DKFZ/nnUNet/files/11350889/side_assert.error2.txt)

Thank you for any help in advance.

Regards,

Gergő

**Comment by gergo7633** [comment]
A newer error with CUDA_LAUNCH_BLOCKING=1 variable:

[side_assert error4.txt](https://github.com/MIC-DKFZ/nnUNet/files/11362310/side_assert.error4.txt)

**Comment by FabianIsensee** [comment]
That's a tough nut to crack. The problem is that I don't see this error on our end which makes it really difficult to debug. 
The error happens here: 

>   File "/home/kutato/nnunet/nnUNet/nnunetv2/training/loss/dice.py", line 95, in forward
>     y_onehot.scatter_(1, gt, 1)

Essentially, there must be a value in gt that doesn't belong there. It seems unlikely that nnunet was able to pass >900 epochs without ever encountering that problem. So this could also be a memory instability on the GPU. Do you have other hardware where you could run this on to see if the error happens there as well?
If possible, you can also share your dataset and I can take a closer look here

**Comment by gergo7633** [comment]
Dear Fabian,

I'm running it with an RTX4080 with reasonable (80-90 sec) epoch times. The other hardware that I can use is an old Tesla K80, with epoch times around 2500 sec (in V1, but it would be similar in V2), which would not be practical. 

Sharing the dataset would be better. I'll send the link to your email address in a minute.

Thankx

**Comment by FabianIsensee** [comment]
Training is running. So far no hiccups
![progress](https://user-images.githubusercontent.com/11302164/235882659-1b23066b-b14c-42c2-83a6-74ff3c001c29.png)
I'll let this run til the end

**Comment by gergo7633** [comment]
Only 2 of the 5 folds have failed in the shared dataset. 

In the following process however, one fold have failed 3 times before it succeeded. Failed at 994, 952 and 970. 


![progress](https://user-images.githubusercontent.com/132046930/235884569-240907a0-d76c-423c-b72a-bc2f9d47b647.png)

Nevertheless, your epoch time is impressive, I have a CPU bottleneck at the moment, but don't think that the RTX4080 can do better than 60sec epoch time.

**Comment by FabianIsensee** [comment]
Which fold fails? I am training fold 0 at the moment

The epoch times are with an A100 GPU + torch.compile ;-)

**Comment by gergo7633** [comment]
The A100 (40GB) is around 8-10 times more expensive as compared to the RTX4080, so the 20sec epoch time is understandable now. 

Fold 3 failed several times. 

Here are my splits to reproduce the fold 3 correctly:

[splits_final.json.txt](https://github.com/MIC-DKFZ/nnUNet/files/11381639/splits_final.json.txt)

**Comment by FabianIsensee** [comment]
splits are seeded so we should get the same :-) I will run fold 3 as well

**Comment by gergo7633** [comment]
Ok, good to know. I thought it is randomly determined during plan_and_preprocess.

**Comment by FabianIsensee** [comment]
So far all folds are still alive

**Comment by gergo7633** [comment]
What an odd feeling, hoping that my data won't work for you.

I usually wish for the opposite.

**Comment by gergo7633** [comment]
I ran cuda-memetest for 6 hours straight. Did over 1600 cycles without a single error. It seems that the VRAM is OK. 



![2023-05-07 13_51_20-Greenshot](https://user-images.githubusercontent.com/132046930/236694420-64fddab3-dfc8-4f4a-abab-be6468e15899.png)

Also interesting that I used the same dataset without BET and completed all 5 folds without error, although I might have been simply lucky this time. The resulting model is considerably better as compared to the dataset with brain extraction. 

Next step is to use an official nvidia docker container and see what happens.

**Comment by lisakausch** [comment]
Hi @gergo7633 ,

could you solve your issue?

Best,
Lisa

**Comment by gergo7633** [comment]
Dear Lisa,


Sort of. I replaced the CPU from Ryzen 5 5600G to 5700X and the issue was
gone.

I don't really know why or how, but it was solved.

Maybe the CPU bottleneck causes some sort of CUDA error that was mitigated
by reducing the bottleneck.

The RTX4080 is now around 10-15% faster so there was a significant CPU
bottleneck.


Regards,

Gergő

lisakausch ***@***.***> ezt írta (időpont: 2023. okt. 25.,
Sze 10:32):
